doc/vrrp/VRRP_Adaptation_HLD.md

```diff
@@ -0,0 +1,712 @@
+ 3) "SAI_ROUTER_INTERFACE_ATTR_SRC_MAC_ADDRESS"
+ 4) "00:00:5E:00:01:06"
+ 5) "SAI_ROUTER_INTERFACE_ATTR_TYPE"
+ 6) "SAI_ROUTER_INTERFACE_TYPE_PORT"
+ 7) "SAI_ROUTER_INTERFACE_ATTR_PORT_ID"
+ 8) "oid:0x1000000000025"
+ 9) "SAI_ROUTER_INTERFACE_ATTR_IS_VIRTUAL"
+10) "true"
+```
+
+#### Modules Design and Flows
+
+##### ADD/DEL VRRP instance
+
+![VRRP ADD/DEL VRRP instance Flows](images/VRRP_Config_Instance_Flow.png)
+
+vrrpcfgd:
+  - Listens to VRRP create, delete and parameter change in CONFIG DB
+  - Update changes by using FRR CLI commands
+
+vrrpd:
+  - Listens to VRRP create, delete and parameter change in vrrpcfgd
+  - Run VRRP state machine
+    - For VRRP instances state transition to Master, set the Macvlan device into protodown off
+    - For VRRP instances state transition to non-Master, set the Macvlan interface into protodown on
+
+vrrpmgrd:
+  - Listens to VRRP create, delete and parameter change in CONFIG DB
+  - Upon change
+    - Add/del VRRP instance corresponding Macvlan device to kernel with IPs and state. 
+    - Updates VRRP instance configuration to the APPL DB, such as interface name, vrid, and IP address type.
+
+vrrpsyncd:
+  - Listens to Macvlan interface programming in kernel.
+  - Update the kernel Macvlan device’s state to the VRRP table entry of APPL DB.
+
+vrrporch:
+  - Listens to VRRP_Table in APP_DB and adds/dels virtual MAC entries in ASIC_DB for VRRP instances
+
+##### Uplink interface tracking
+
+![VRRP Tracking interface Flows](images/VRRP_Tracking_Interface_Flow.png)
+
+vrrpsyncd:
+  - Listen for interfaces state change events notified by the kernel. 
+  - Match VRRP instance tracking interface and recalculate priority
+    - If track interface goes down, the priority of VRRP instance will be reduced by weight
+    - If track interface goes up, the priority of VRRP instance will be added by weight
+  - Update priority change by using FRR CLI commands
+
+vrrpd:
+  - Update VRRP packet priority field
+  - Send VRRP packet to Linux Network stack
+
+### SAI API 
+
+VRRP virtual router interface can be created by setting the attribute “SAI_ROUTER_INTERFACE_ATTR_IS_VIRTUAL” with create_router_interface API.
+```
+ * @brief RIF creation is a virtual RIF.
+ *
+ * Create a Virtual RIF object, which only programs the ingress router MAC.
+ * This simplifies the management of VRRP master router's configuration in
+ * SAI adapter, as defined by RFC 5798 (or similar proprietary protocols).
+ * Using a Virtual RIF allows SAI to optimize resources, so neighbor entries
+ * cannot be learned on a Virtual RIF. On a virtual RIF following attributes
+ * are invalid: ADMIN state, MTU size, packet action and multicast enable.
+ * Alternatively VRRP can also be configured using native RIF objects without
+ * using VIRTUAL attribute, with the expectation that SAI adapter will consume
+ * resources that will not be used.
+ *
+ * @type bool
+ * @flags CREATE_ONLY
+ * @default false
+ */
+
+SAI_ROUTER_INTERFACE_ATTR_IS_VIRTUAL,
```

Comment:
please describe in hld if this attributed is not supported by a particular vendor sai implementation.

Reply:
ok